sdk/core/core-client/src/serviceClient.ts

```diff
@@ -190,12 +191,26 @@ export class ServiceClient {
         options.onResponse(rawResponse, flatResponse);
       }
       return flatResponse;
-    } catch (error) {
-      if (error.response) {
-        error.details = flattenResponse(
-          error.response,
-          operationSpec.responses[error.statusCode] || operationSpec.responses["default"]
-        );
+    } catch (e) {
+      const error: any = e;
+      if (typeof error === "object" && error) {
+        let rawResponse: PipelineResponse = {
+          request,
+          headers: createHttpHeaders(),
+          status: 0,
+        };
+        let flatResponse: unknown;
+        if (error.response) {
+          rawResponse = error.response;
+          flatResponse = flattenResponse(
+            rawResponse,
+            operationSpec.responses[error.statusCode] || operationSpec.responses["default"]
+          );
+          error.details = flatResponse;
+        }
+        if (options?.onResponse) {
+          options.onResponse(rawResponse, flatResponse, error);
+        }
```

Comment:
open question: should we move this block up into the `if(error.response)` so it only triggers on RestErrors that have a valid response?

Reply:
Could there be a rest error that does not have a response property? if not, then I think it makes sense it move it up.

Reply:
Not a RestError per se, but in theory a policy could throw whatever it wanted and it would end up here (e.g. RangeError.) 

I think I'm with you that the nature of the callback is best effort when we have a response, not to bulletproof a try/catch

Reply:
Yeah I agree. I think when a customer uses this, they would consider a response as something coming from the service they're communicating with so giving them an error that happened before receiving a response could be surprising to them.